### PR TITLE
[WIP] Refactor LazyTensor unit tests

### PR DESCRIPTION
**Pull request by gpleiss**
We've seen a number of bugs recently in LazyTensor behaviors. Some LTs have comprehensive unit tests, covering lots of edge cases. Others do not.

I propose creating super classes for all LT unit tests that contain lots of test cases. For each LT unit test, we will only have to write a method that constructs/evaluates the LT.

I started on a couple LTs.

### CODE REVIEW

File: gpytorch/lazy/constant_mul_lazy_tensor.py [hunk 1]
```diff
@@ -55,50 +55,50 @@ class ConstantMulLazyTensor(LazyTensor):
...
                         "Got a %dD lazy var. with size %s"
-                        % (numel, numel, lazy_var.ndimension(), repr(lazy_var.size()))
+                        % (numel, numel, base_lazy_tensor.ndimension(), repr(base_lazy_tensor.size()))
                     )
 
             elif constant.numel() == 1:
                 constant = constant.squeeze()
         else:
-            constant = torch.tensor(constant, device=lazy_var.device, dtype=torch.float32)
+            constant = torch.tensor(constant, device=base_lazy_tensor.device, dtype=torch.float32)
```
Comment (Balandat) [inline gpytorch/lazy/constant_mul_lazy_tensor.py]: seems like this should be using the dtype of the base_lazy_tensor as well

File: test/lazy/_lazy_tensor_test_case.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+    def test_matmul_vec(self):
+        lazy_tensor = self.create_lazy_tensor()
+        lazy_tensor_copy = lazy_tensor.clone()
+        evaluated = self.evaluate_lazy_tensor(lazy_tensor_copy)
+
+        test_vector = torch.randn(lazy_tensor.size(-1))
+        res = lazy_tensor.matmul(test_vector)
+        actual = evaluated.matmul(test_vector)
+        assert(approx_equal(res, actual))
```
Comment (Balandat) [inline test/lazy/_lazy_tensor_test_case.py]: lets make this a `self.assertTrue(approx_equal(res, actual))` (and similarly in other places)

**Comment by gpleiss** [comment]
This PR is now ready to be merged. There's a couple of bug fixes:

- Fix logdet calculations for batch LazyTensors
- Fix __getitem__ for KroneckerProductLazyTensors

**Comment by Balandat** [comment]
@gpleiss, what about my comment on hard-coding the dtype to `float32` in `constant_mul_lazy_tensor.py`? Looks like that would break things when running as double?

**Comment by gpleiss** [comment]
@Balandat sorry I missed that! Fixing it now

File: gpytorch/lazy/constant_mul_lazy_tensor.py (hunk 1 above)
Comment (gpleiss) [reply gpytorch/lazy/constant_mul_lazy_tensor.py]: Updated

**Comment by Balandat** [comment]
I'm seeing a test failure in `TestMulLazyTensorMulti`, not sure why this didn't show up in Travis. 

Also, can we add cuda versions for all the LazyTensor test cases? We've had plenty of bugs due to device incompatibilities that would be easy to avoid with tests. Should be simple to do by following the pattern I used e.g. in `test_lkj_prior.py`

**Comment by gpleiss** [comment]
I'll take a look at the test failure and open another PR for the cuda tests

**Comment by gpleiss** [comment]
Test failure should be fixed now. Adding an issue for cuda tests.
